go/tools/bzltestutil/lcov.go

```diff
@@ -0,0 +1,144 @@
+	"fmt"
+	"io"
+	"os"
+	"regexp"
+	"sort"
+	"strconv"
+	"strings"
+)
+
+// ConvertCoverToLcov converts the go coverprofile file coverage.dat.cover to
+// the lcov format and stores it in coverage.dat, where it is picked up by
+// Bazel.
+// The conversion emits line and branch coverage, but not function coverage.
+func ConvertCoverToLcov() error {
+	// The value of test.coverprofile has been set to
+	// ${COVERAGE_OUTPUT_PATH}.cover by the generated TestMain. We have to strip
+	// the ".cover" suffix here so that the generated lcov report is at the
+	// location where Bazel's lcov merge picks it up.
+	inPath := flag.Lookup("test.coverprofile").Value.String()
+	outPath := strings.TrimSuffix(inPath, ".cover")
+
+	in, err := os.Open(inPath)
+	if err != nil {
+		// This can happen if there are no tests and should not be an error.
+		return nil
+	}
+	defer in.Close()
+
+	out, err := os.Create(outPath)
+	if err != nil {
+		return err
+	}
+	defer out.Close()
+
+	return convertCoverToLcov(bufio.NewScanner(in), out)
+}
+
+var coverLinePattern = regexp.MustCompile(`^(?P<path>.*):(?P<startLine>\d+)\.(?P<startColumn>\d+),(?P<endLine>\d+)\.(?P<endColumn>\d+) (?P<numStmt>\d+) (?P<count>\d+)$`)
+const pathIdx = 1
+const startLineIdx = 2
+const endLineIdx = 4
+const countIdx = 7
+
+func convertCoverToLcov(cover *bufio.Scanner, lcov io.StringWriter) error {
+	currentPath := ""
+	var lineCounts map[uint32]uint32
+	for cover.Scan() {
+		l := cover.Text()
+		m := coverLinePattern.FindStringSubmatch(l)
+		if m == nil {
+			if strings.HasPrefix(l, "mode: ") {
+				continue
+			}
+			return fmt.Errorf("invalid go cover line: %s", l)
+		}
+
+		if m[pathIdx] != currentPath {
+			if currentPath != "" {
+				if err := emitLcovLines(lcov, currentPath, lineCounts); err != nil {
+					return err
+				}
+			}
+			currentPath = m[pathIdx]
+			lineCounts = make(map[uint32]uint32)
+		}
+
+		startLine, err := strconv.ParseUint(m[startLineIdx], 10, 32)
+		if err != nil {
+			return err
+		}
+		endLine, err := strconv.ParseUint(m[endLineIdx], 10, 32)
+		if err != nil {
+			return err
+		}
+		count, err := strconv.ParseUint(m[countIdx], 10, 32)
+		if err != nil {
+			return err
+		}
+		for line := uint32(startLine); line <= uint32(endLine); line++ {
+			prevCount, ok := lineCounts[line]
+			if !ok || uint32(count) > prevCount {
+				lineCounts[line] = uint32(count)
+			}
+		}
+	}
+	if currentPath != "" {
+		if err := emitLcovLines(lcov, currentPath, lineCounts); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func emitLcovLines(lcov io.StringWriter, path string, lineCounts map[uint32]uint32) error {
+	_, err := lcov.WriteString(fmt.Sprintf("SF:%s\n", path))
+	if err != nil {
+		return err
+	}
+
+	// Emit the coverage counters for the individual source lines.
+	sortedLines := make([]uint32, len(lineCounts))
+	i := 0
+	for line := range lineCounts {
+		sortedLines[i] = line
+		i++
+	}
+	sort.Slice(sortedLines, func(i, j int) bool { return sortedLines[i] < sortedLines[j] })
```

Comment:
```suggestion
	sort.Ints(sortedLines)
```

Reply:
The line numbers are parsed from the go cover report, which internally uses `uint32` to represent them. I wouldn't expect source files with more than INT_MAX lines to be common, but it does feel a bit unclean to me to use a type that may not be able to faithfully represent all values supported by the underlying protocol. But if you value code simplicity more here, I can switch to `int`.